src/helpers/pollUpdates.ts

```diff
@@ -1,38 +1,78 @@
 import { updateAvailable } from '../app/notifications/actions';
 import { Store } from '../app/types';
+import { assertDocument } from '../app/utils';
 import { APP_ENV, RELEASE_ID } from '../config';
 
+/*
+ * when a page is initially loaded, the built in release
+ * id might legitimately be different from the environment.json
+ * release id if a deployment is currently in progress. in order
+ * to avoid repeatedly prompting to reload in this case, we wait
+ * to actually observe the environment.json release id change
+ * before prompting.
+ *
+ * there is potential to miss the environment.json changing if
+ * the window doesn't have focus, so after a bit we can start
+ * trusting the release id from environment.json and just compare
+ * it to our built one
+ */
+
+const pollInterval = 1000 * 60; // 1m
+const trustAfter = 1000 * 60 * 60; // 1h
+const pageLoaded = new Date().getTime();
+const trustRelease = () => (new Date().getTime()) - pageLoaded > trustAfter;
+let previousObservedReleaseId: string | undefined;
+
 export type Cancel = () => void;
+interface Environment {
+  release_id: string;
+}
+
+const processEnvironment = (store: Store, environment: Environment) => {
+  const releaseId = environment.release_id;
+
+  if (
+    environment.release_id !== RELEASE_ID
+    && (trustRelease() || previousObservedReleaseId !== releaseId)
+  ) {
+    store.dispatch(updateAvailable());
+  }
 
-export const poll = (store: Store, cancel: Cancel) => async() => {
+  previousObservedReleaseId = releaseId;
+};
+
+export const poll = (store: Store) => async() => {
   const environment = await fetch('/rex/environment.json')
-    .then((response) => response.json())
-    .catch(() => ({release_id: RELEASE_ID}))
+    .then((response) => response.json() as Promise<Environment>)
+    .catch(() => null)
   ;
```

Comment:
Seems strange to me to call `.then` & `.catch` on a function that you're also using `await` on.  I'm guessing that the then will be called and then the await ?

Reply:
yea it'll await the final promise, easier to chain the call to `json` and more terse than a real try/catch